aiogram/dispatcher/requirement.py

```diff
@@ -0,0 +1,173 @@
+                await req.initialize_children(data, cache_dict, stack)
+
+                if req.use_cache and req.cache_key in cache_dict:
+                    data[req_id] = cache_dict[req.cache_key]
+
+                else:
+                    data[req_id] = await initialize_callable_requirement(req, data, stack)
+
+                if req.use_cache:
+                    cache_dict[req.cache_key] = data[req_id]
+
+    async def __call__(
+        self, cache_dict: Dict[CacheKeyType, Any], stack: AsyncExitStack, data: Dict[str, Any],
+    ) -> T:
+        await self.initialize_children(data, cache_dict, stack)
+
+        if self.use_cache and self.cache_key in cache_dict:
+            return cast(T, cache_dict[self.cache_key])
+        else:
+            result = await initialize_callable_requirement(self, data, stack)
+            if self.use_cache:
+                cache_dict[self.cache_key] = result
+            return result
+
+
+class CachedRequirement(Requirement[T]):
+    __slots__ = "cache_key", "value_on_miss"
+
+    def __init__(self, cache_key: CacheKeyType, value_on_miss: T):
+        self.cache_key: CacheKeyType = cache_key
+        self.value_on_miss = value_on_miss
+
+    async def __call__(
+        self, cache_dict: Dict[CacheKeyType, Any], stack: AsyncExitStack, data: Dict[str, Any],
+    ) -> T:
+        return cache_dict.get(self.cache_key, self.value_on_miss)
+
+
+async def initialize_callable_requirement(
+    required: CallableRequirement[T], data: Dict[str, Any], stack: AsyncExitStack
+) -> T:
+    actual_data = required.filter_kwargs(data)
+    async_cm: Optional[Any] = None
+
+    if required.generator_type is GeneratorKind.async_gen:
+        async_cm = asynccontextmanager(required.callable)(**actual_data)  # type: ignore
+    elif required.generator_type is GeneratorKind.plain_gen:
+        async_cm = move_to_async_gen(contextmanager(required.callable)(**actual_data))  # type: ignore
+
+    if async_cm is not None:
+        return await stack.enter_async_context(async_cm)
+    else:
+        result = required.callable(**actual_data)
+        if isinstance(result, Awaitable):
+            return cast(T, await result)
+        return cast(T, result)
+
+
+def get_reqs_from_callable(callable_: _RequiredCallback[T]) -> Dict[str, Requirement[Any]]:
+    signature = inspect.signature(callable_)
+    return {
+        param_name: param_self.default
+        for param_name, param_self in signature.parameters.items()
+        if isinstance(param_self.default, Requirement)
+    }
+
+
+def get_reqs_from_class(cls: Type[Any]) -> Dict[str, Requirement[Any]]:
+    return {
+        req_attr: req for req_attr, req in cls.__dict__.items() if isinstance(req, Requirement)
+    }
+
+
+def require(
+    what: _RequiredCallback[T],
+    *,
+    cache_key: Optional[CacheKeyType] = None,
+    use_cache: bool = True,
+) -> T:
+    return CallableRequirement(what, cache_key=cache_key, use_cache=use_cache)  # type: ignore
```

Comment:
Why this factory is needed? Why `Requirement(...)` can't be used directly?

Reply:
I wanted to differentiate requirements. I clearly need to rethink that now.